File: frontend/src/types/ptp.ts
/**
 * TypeScript types for PTP (Profile to Print) feature
 * Matches backend Pydantic models in models/race.py and models/ptp.py
 */

export type RavitoType = 'eau' | 'bouffe' | 'assistance';

export interface RaceAidStation {
  id?: string;
  name: string;
  distance_km: number;
  elevation?: number;
  type: RavitoType;
  services?: string[];
  cutoff_time?: string;
  position_order: number;
}

export interface Race {
  id: string;
  name: string;
  slug: string;
  description?: string;
  gpx_content: string;
  total_distance_km?: number;
  total_elevation_gain?: number;
  total_elevation_loss?: number;
  start_location_lat?: number;
  start_location_lon?: number;
  is_published: boolean;
  aid_stations: RaceAidStation[];
  created_at: string;
  updated_at: string;
}

export interface RaceCreate {
  name: string;
  slug: string;
  description?: string;
  gpx_content: string;
  is_published?: boolean;
  aid_stations?: Omit<RaceAidStation, 'id'>[];
}

export interface RaceUpdate {
  name?: string;
  slug?: string;
  description?: string;
  gpx_content?: string;
  is_published?: boolean;
  aid_stations?: Omit<RaceAidStation, 'id'>[];
}

export interface RaceListResponse {
  races: Race[];
  total: number;
}

export interface AdminLoginRequest {
  password: string;
}

export interface AdminLoginResponse {
  success: boolean;
  token: string;
  message: string;
}

export interface ParsedRavito {
  name: string;
  distance_km: number;
  elevation?: number;
  type: RavitoType;
  services?: string[];
  cutoff_time?: string;
}

export interface ParsedRavitoTable {
  ravitos: ParsedRavito[];
  warnings?: string[];
}

export interface SunTimes {
  sunrise: string;
  sunset: string;
  solar_noon: string;
  day_length: string;
  civil_twilight_begin: string;
  civil_twilight_end: string;
}

export interface GetSunTimesRequest {
  lat: number;
  lon: number;
  date: string; // YYYY-MM-DD
}

export interface GetSunTimesResponse {
  success: boolean;
  data?: SunTimes;
  error?: string;
}

// Runner configuration for roadbook
export interface RunnerConfig {
  departure_time: string; // ISO datetime
  flask_capacity: number; // 1, 2, or 3 flasks
  pace_override?: number; // km/h, optional override of Naismith
  notes: Record<string, string>; // aid_station_id -> note
}

// Computed passage time for display on profile
export interface PassageTime {
  aid_station_id: string;
  distance_km: number;
  estimated_arrival: string; // ISO datetime
  estimated_departure?: string;
  time_from_start_minutes: number;
}

File: backend/app/models/ptp.py
"""
PTP (Profile to Print) models for API
"""
from typing import List, Optional
from pydantic import BaseModel
from app.models.race import RavitoType


class ParsedRavito(BaseModel):
    """Ravito parsed from external table"""
    name: str
    distance_km: float
    elevation: Optional[int] = None
    type: Optional[RavitoType] = None
    services: Optional[List[str]] = None
    cutoff_time: Optional[str] = None


class ParsedRavitoTable(BaseModel):
    """Result of parsing external ravito table"""
    ravitos: List[ParsedRavito]
    race_name: Optional[str] = None
    total_distance: Optional[float] = None


class ParseRavitoTableRequest(BaseModel):
    """Request to parse ravito table with Claude"""
    raw_text: str


class ParseRavitoTableResponse(BaseModel):
    """Response from parsing ravito table"""
    success: bool
    data: Optional[ParsedRavitoTable] = None
    error: Optional[str] = None


class SunTimes(BaseModel):
    """Sunrise/sunset times for a specific date"""
    sunrise: str  # ISO datetime
    sunset: str
    civil_twilight_begin: str
    civil_twilight_end: str
    date: str


class GetSunTimesRequest(BaseModel):
    """Request for sun times"""
    lat: float
    lon: float
    dates: List[str]  # ISO date strings YYYY-MM-DD


class GetSunTimesResponse(BaseModel):
    """Response with sun times"""
    success: bool
    sun_times: List[SunTimes] = []
    error: Optional[str] = None
